__tests__/fetch_request.js

```diff
@@ -0,0 +1,234 @@
+      .toStrictEqual({ ...defaultHeaders, custom: "Header", "Content-Type": "application/json", "test": "header"})
+  })    
+
+  test('headers win over contentType', () => {
+    const request = new FetchRequest("get", "localhost", { contentType: "application/json", headers: { "Content-Type": "this/overwrites" } })
+    expect(request.fetchOptions.headers)
+      .toStrictEqual({ ...defaultHeaders, "Content-Type": "this/overwrites"})
+  })    
+
+  test('serializes JSON to String', () => {
+    const jsonBody = { some: "json" }
+    let request
+    request = new FetchRequest("get", "localhost", { body: jsonBody, contentType: "application/json" })
+    expect(request.fetchOptions.body).toBe(JSON.stringify(jsonBody))
+
+    request = new FetchRequest("get", "localhost", { body: jsonBody })
+    expect(request.fetchOptions.body).toBe(JSON.stringify(jsonBody))
+  })
+
+  test('not serializes JSON with explicit different contentTyp', () => {
+    const jsonBody = { some: "json" }
+    const request = new FetchRequest("get", "localhost", { body: jsonBody, contentType: "not/json" })
+    expect(request.fetchOptions.body).toBe(jsonBody)
+  })
+
+  test('set redirect', () => {
+    let request
+    const redirectTypes = [ "follow", "error", "manual" ]
+    for (const redirect of redirectTypes) {
+      request = new FetchRequest("get", "localhost", { redirect })
+      expect(request.fetchOptions.redirect).toBe(redirect)
+    }
+    
+    request = new FetchRequest("get", "localhost")
+    expect(request.fetchOptions.redirect).toBe("follow")
+
+    // maybe check for valid values and default to follow?
+    // https://developer.mozilla.org/en-US/docs/Web/API/Request/redirect
+    request = new FetchRequest("get", "localhost", { redirect: "nonsense"})
+    expect(request.fetchOptions.redirect).toBe("nonsense")
+  })
+
+  test('sets signal', () => {
+    let request
+    request = new FetchRequest("get", "localhost")
+    expect(request.fetchOptions.signal).toBe(undefined)
+
+    request = new FetchRequest("get", "localhost", { signal: "signal"})
+    expect(request.fetchOptions.signal).toBe("signal")
+  })
+
+  test('has fixed credentials setting which cannot be changed', () => {
+    let request
+    request = new FetchRequest("get", "localhost")
+    expect(request.fetchOptions.credentials).toBe('same-origin')
+
+    // has no effect
+    request = new FetchRequest("get", "localhost", { credentials: "omit"})
+    expect(request.fetchOptions.credentials).toBe('same-origin')
+  })  
+})
+
+describe('query params are parsed', () => {
+  test('anchors are rejected', () => {
+    const testRequest = new FetchRequest("post", "localhost/test?a=1&b=2#anchor", { query: { c: 3 } })
+    expect(testRequest.url).toBe("localhost/test?a=1&b=2&c=3")
+    // const brokenRequest = new FetchRequest("post", "localhost/test#anchor", { query: { a: 1, b: 2, c: 3 } })
+    // expect(brokenRequest.url).toBe("localhost/test?a=1&b=2&c=3")
```

Comment:
Is there a reason to keep those lines?

Reply:
Yes, there is! This test-case will fail. expected is `localhost/test?a=1&b=2&c=3`, but received is `localhost/test#anchor?a=1&b=2&c=3`.

I'd propose one of these solutions.
1. I could fix the functionality in this PR (would somehow mix stuff up)
2. I do another PR with the fix and uncomment the case (clean separation which PR does what)

I actually don't care, but I'm not the one pushing approve xD

Reply:
I think that's better we go with the second option after this PR gets merged 👍